northstar/src/runtime/pipe.rs

```diff
@@ -193,63 +276,233 @@ mod tests {
     fn read_write() {
-        let (mut r, mut w) = pipe().unwrap();
+        let (mut read, mut write) = pipe().unwrap();
 
-        let w_task = std::thread::spawn(move || {
+        let writer = thread::spawn(move || {
             for n in 0..=65535u32 {
-                w.write(&n.to_be_bytes()).unwrap();
+                write.write(&n.to_be_bytes()).unwrap();
             }
         });
 
-        let r_task = std::thread::spawn(move || {
-            let mut buf = [0u8; 4];
-            for n in 0..=65535u32 {
-                r.read_exact(&mut buf).unwrap();
-                assert_eq!(buf, n.to_be_bytes());
-            }
-        });
+        let mut buf = [0u8; 4];
+        for n in 0..=65535u32 {
+            read.read_exact(&mut buf).unwrap();
+            assert_eq!(buf, n.to_be_bytes());
+        }
 
-        w_task.join().unwrap();
-        r_task.join().unwrap();
+        writer.join().unwrap();
     }
 
     #[tokio::test]
-    async fn async_read_write() {
-        let (r, w) = pipe().unwrap();
+    /// Test async version of read and write
+    async fn r#async() {
+        let (read, write) = pipe().unwrap();
 
-        let mut async_reader: AsyncPipeReader = r.try_into().unwrap();
-        let mut async_writer: AsyncPipeWriter = w.try_into().unwrap();
+        let mut read: AsyncPipeRead = read.try_into().unwrap();
+        let mut write: AsyncPipeWrite = write.try_into().unwrap();
 
-        let w_task = tokio::spawn(async move {
+        let write = tokio::spawn(async move {
             for n in 0..=65535u32 {
-                async_writer.write(&n.to_be_bytes()).await.unwrap();
+                write.write(&n.to_be_bytes()).await.unwrap();
             }
         });
 
-        let r_task = tokio::spawn(async move {
-            let mut buf = [0u8; 4];
-            for n in 0..=65535u32 {
-                async_reader.read_exact(&mut buf).await.unwrap();
-                assert_eq!(buf, n.to_be_bytes());
+        let mut buf = [0u8; 4];
+        for n in 0..=65535u32 {
+            read.read_exact(&mut buf).await.unwrap();
+            assert_eq!(buf, n.to_be_bytes());
+        }
+
+        write.await.unwrap()
+    }
+
+    #[test]
+    /// Fork test
+    fn fork() {
+        let (mut read, mut write) = pipe().unwrap();
+
+        match unsafe { unistd::fork().unwrap() } {
+            unistd::ForkResult::Parent { child } => {
+                drop(read);
+                for n in 0..=65535u32 {
+                    write.write(&n.to_be_bytes()).unwrap();
+                }
+                nix::sys::wait::waitpid(child, None).ok();
             }
-        });
-        tokio::try_join!(w_task, r_task).unwrap();
+            unistd::ForkResult::Child => {
+                drop(write);
+                let mut buf = [0u8; 4];
+                for n in 0..=65535u32 {
+                    read.read_exact(&mut buf).unwrap();
+                    assert_eq!(buf, n.to_be_bytes());
+                }
+            }
```

Comment:
I added the fork tests because `tokio` in a forked child doesn't work. This is known issue. I just wanted to ensure that the fork doesn't interfere with the "normal" blocking code. 

The rust test framework starts each test in a thread with a filter. So also the child will exit after the test and not continue to execute something. 

I'll add explicit an `exit` anyways for clearness.

Reply:
I don't think the test framework is protected against `fork`. I just made a small test:

```rust
#[test]
fn fork_test() {
    match unsafe { nix::unistd::fork() }.unwrap() {
        nix::unistd::ForkResult::Parent { .. } => {}
        nix::unistd::ForkResult::Child => {}
    }
}

#[test]
fn hello() {
    println!("hello, i'm {}", std::process::id());
}
```
and got this as output:

```
running 2 tests
test fork_test ... ok
test hello ... hello, i'm 1869154
ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

ok
test hello ... hello, i'm 1869155
ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

```

i'm using the `--nocapture` flag to get the `stdout` from both processes.

Reply:
Thanks for checking. Didn't know that. With the `exit` it should be fine.